### PR TITLE
Reduce provisioner work

### PR DESCRIPTION
**Pull request by alexgparity**
Companion of https://github.com/paritytech/substrate/pull/12749  

Cumulus companion: https://github.com/paritytech/cumulus/pull/1940

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2073098

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2074051

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2074425

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2074537

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2074870

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2076756

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2077107

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [inline node/core/parachains-inherent/src/lib.rs:148]: This should not be created.

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [inline node/core/parachains-inherent/src/lib.rs:54]: Most of the logic in this function should be moved into the `provide_inherent_data` function.

File: node/core/parachains-inherent/src/lib.rs
Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:148]: Why not? To save a few bytes and allocation?

File: node/core/parachains-inherent/src/lib.rs
Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:54]: Why so? I like how this is actually a separate function handing all the data gathering.

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [reply node/core/parachains-inherent/src/lib.rs:148]: No. This type just doesn't make any sense. We already have the `InherentDataProvider`. The implementation just needs to be changed. Here you created a second type and hacked the old type into this.

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [reply node/core/parachains-inherent/src/lib.rs:54]: You can still put it into a separate function. I just don't like the duplication of the InherentDataProvider. This doesn't make any sense to me.

File: node/core/parachains-inherent/src/lib.rs
Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:148]: Fair. I was thinking of calling this the `ParachainsInherentDataProviderCreator` but I get where you are coming from.

Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:148]: The only reason to keep the old instance and create a new one as a "provider creator" in the first place was the benchmarking code here https://github.com/paritytech/polkadot/pull/6328/files#diff-2e324183f8b01329c246bc1cf919b043150b5a88c94bf80d989a4034c0b2bd8dR372-R375. As `provide_inherent_data` assumes all the data comes from `self`, I'm having a bit of a struggle here.

Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:148]: Do I create a `BenchmarkProvider` that always returns that mocked data?

Comment (bkchr) [reply node/core/parachains-inherent/src/lib.rs:148]: Either that or make the provider an enum. Where one variant returns the given data.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2089899

File: node/client/src/benchmarking.rs
Comment (bkchr) [inline node/client/src/benchmarking.rs:378]: This is wrong. You need to use the correct `ParachainsInherentDataProvider`. And if you are re-implementing the `InherentDataProvider` any way, you can don't need to do it. Just create the `InherentData` manually.

Comment (alexgparity) [reply node/client/src/benchmarking.rs:378]: I'm not following. I asked if I should do what I did https://github.com/paritytech/polkadot/pull/6328#discussion_r1032602742 and you said I can. What was my misunderstanding?

At the end of the day, what exactly is this supposed to benchmark? `provide_inherent_data(...)` for `ParachainsInherentDataProvider`? `dst_inherent_data.put_data(...)` for the default `ParachainsInherentData`?

Comment (bkchr) [reply node/client/src/benchmarking.rs:378]: The first parameter to `put_data` is a unique identifier. If you change this identifier, the runtime functionality will not find this inherent data.

However, I just checked the code and we don't really need this.  @ggwpez can chime in to support my assumption here, but we probably use this to build some empty block for benchmarking. This empty block would also work without providing any inherent data.

TLDR, you should be able to remove the entire code here.

Comment (ggwpez) [reply node/client/src/benchmarking.rs:378]: I think the `paras-inheret` pallet requires `PARACHAINS_INHERENT_IDENTIFIER` to be present, but it can be anything.  
So something like this should be enough:  
```rust
inherent_data.put_data(polkadot_primitives::v2::PARACHAINS_INHERENT_IDENTIFIER, &para_data);
```

I cannot compile the code even after diener but you can test it with `benchmark overhead --dev --max-ext-per-block 10` sub-command.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2100898

File: node/client/src/benchmarking.rs
Comment (bkchr) [inline node/client/src/benchmarking.rs:373]:
```suggestion
		.put_data(polkadot_primitives::v2::PARACHAINS_INHERENT_IDENTIFIER, &para_data)?;
```

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [inline node/core/parachains-inherent/src/lib.rs:57]:
```suggestion
	async fn create(&self
```

File: node/core/parachains-inherent/src/lib.rs
Comment (bkchr) [inline node/core/parachains-inherent/src/lib.rs:141]:
```suggestion
		let inherent_data = self.create()
		.await
		.map_err(|e| sp_inherents::Error::Application(Box::new(e)))?;
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2105692

File: node/core/parachains-inherent/src/lib.rs
Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:57]: Revert as this would need `self`to be `mut` which in the end would need `self` in `InherentDataProvider` to be mut which feels wrong to change for this one case.

File: node/core/parachains-inherent/src/lib.rs
Comment (alexgparity) [reply node/core/parachains-inherent/src/lib.rs:141]: Revert as this would need `self`to be `mut` which in the end would need `self` in `InherentDataProvider` to be mut which feels wrong to change for this one case.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2105797

**Comment by ggwpez** [comment]
Please always merge with `bot merge` since this is now blocking the CI for all other Polkadot pull requests.  
Eg: https://github.com/paritytech/polkadot/pull/6376

**Review by eskimor — APPROVED** [review]

**Comment by ggwpez** [comment]
It also needs a Cumulus CI since you merged the Substrate MR while the Cumulus check was red
image

I see that the Polkadot benches check here is red as well (not your fault) I am fixing it.

**Comment by alexgparity** [comment]
> It also needs a Cumulus companion since you merged the Substrate MR while the Cumulus CI check was red image
> 
> I see that the Polkadot benches check here is red as well (not your fault) I am fixing it.

My impression is that jobs marked as "not required" are well, not required to pass? I suppose it is required to pass though?

**Comment by ggwpez** [comment]
> My impression is that jobs marked as "not required" are well, not required to pass? I suppose it is required to pass though?

Yea good Q. I asked in the CI/CD what is the reason for not requiring it.

**Comment by alexgparity** [comment]
@ggwpez Meanwhile, I did https://github.com/paritytech/cumulus/pull/1940. `cargo check` works locally for me when `diener` points to the polkadot companion. No clue what to do on the PR to make it work. Didn't find any contributing doc as there is in substrate. If you could advise, please.

**Comment by ggwpez** [comment]
The process is analogous to a Substrate->Polkadot companion. I added you Cumulus MR as companion here; lets see if the check goes green :crossed_fingers:  
There is the [contributing](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc#updating-polkadot-as-well) doc in Substrate. I dont think there is more docs currently.

**Comment by ggwpez** [comment]
Okay looks good. If you could approve this https://github.com/paritytech/substrate/pull/12829 then we can merge it and get the check here green as well @alexgparity

**Comment by alexgparity** [comment]
@ggwpez done

**Review by ggwpez — APPROVED** [review]
LGTM but @bkchr should approve as well.  
Then we can unblock the CI.

**Pull request merged**
